artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonAbstractReceiver.java

```diff
@@ -346,6 +346,7 @@ public void onMessageComplete(Delivery delivery,
       connection.requireInHandler();
 
       try {
+         this.messageReader.close();
```

Comment:
Can you elaborate the actual issue and how this close moving here makes a difference? Looking at the large message reader, its close method, and how/when the reader calls this onMessageComplete() method, I'm not really seeing how this would alter the behaviour to fix something?

Reply:
@gemmellr leaving the close to happen after the advance would allow a situation where we confirm the client the message has been accepted. Later the server crashed, even after a commit (depending on how fast it happened) and the file wasn't yet closed.. leaving a truncated file in the file system...


After a restart the file still zeroed because the buffer that contained the file before the close is gone.

Reply:
2 days to find this... not a very fun experience..


I thought the BytesMessage was something else leaking to the client.

Reply:
Actually.. I'm wrong...


I found this based on the bisect.. since the problem is the ingestion this is the only part that changed between the commit that introduced this and after...


I'm puzzled now after I read the close method... 


it was the only change between the offending commit and after (on the producer's server's side).

Reply:
That doesn't make a ton of sense though as the onMessageCompleted is call only after the large message release resources has been called meaning the large message should have already been "closed" unless you mean something other than closed when you are saying closed.  Also the close in the try / finally block is a fail safe to ensure that if an error occurs on that path then things actually do get closed, which again should have already been done.  We need a bit more detail on this as currently it simply doesn't seem right as I'd have expected that close call to be a no-op as AMQPLargeMessageReader would check state of 'currentMessage' and see that it was already null as it is nulled before onMessageCompleted is ever called.

Reply:
the test started to faill after c83ed8957d9d7f06bb29c0fce563fe2e3462993e

Reply:
Are you sure the test has truly stopped failing? Perhaps moving it is simply affecting timing as there would be a few volatile reads that could alter minor timing sensitive cases.  Really understanding what is going on here is key because any looking though the code so far leads me to understand that the close() call will simply check and see that the large message was already released and nulled before the code that calls onMessageComplete in AMQPLargeMessageReader::addBytes

Reply:
It did stop failing.  Many repetitions. 


I think there is a failure check where I delete the file.    Moving the close reader before the advance stopped it from happening.

Reply:
I will confirm is the failure check later today (I'm out of my desk at the moment)

Reply:
First thing I would do is confirm that the close call in the onMessageComplete is actually doing anything as by default it would not be doing anything other than toggling the boolean 'closed' as the currentMessage would be null and no interactions with the large message would occur there.

Reply:
@tabish121 @gemmellr the issue is indeed the close after a failure removing the file.


this method here gets closed on a network error, disconnect... etc:

https://github.com/apache/activemq-artemis/blob/ffafbf350af6f92f7984cbc251be8f2d5dfcf2d4/artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonAbstractReceiver.java#L369-L374



If I don't close the reader before the advance I could get to a race where the message was already confirmed and this close could eventually happen.


Closing the reader before the advance fixed the issue.



I will add a big // COMMENT block explaining the situation here.

Reply:
it wasn't very pleasant to find such tiny thing to be honest :)  (just blaming myself.. lol as I made the change)

Reply:
The outcome is a test failure on InterruptLargeMessageTest. This commit https://github.com/apache/activemq-artemis/commit/c83ed8957d9d7f06bb29c0fce563fe2e3462993e  broke the test as it is not setting the AMQPMessageReader as closed when it's supposed. as a consequence dropped connections will make the failureCallbacks to remove a "pending" large message which is not actually pending on this case.

Reply:
the issue is actually the call to actualDelivery ...if we perform actualDelivery before calling messageReader.close() you could have the message being part of the queue while the reader will be marked as open. a connection failure happens and the body of the message removed. as a consequence the client will receive an empty body message which the server is translating as a bytesMessage.

Reply:
200 passes with testInterruptLargeMessageAMQPTXPaging with this change on the reader.close();


reverted the fix... the test failed within 2 retries.